unmarshaller.go

```diff
@@ -1,117 +1,114 @@
-package gocsv
+package commando
 
 import (
-	"encoding/csv"
 	"fmt"
+	"io"
 	"reflect"
 )
 
 // Unmarshaller is a CSV to struct unmarshaller.
 type Unmarshaller struct {
-	reader                 *csv.Reader
-	Headers                []string
-	fieldInfoMap           []*fieldInfo
-	MismatchedHeaders      []string
-	MismatchedStructFields []string
-	outType                reflect.Type
+	config *validConfig
+	line   int
+	reader Reader
 }
 
-// NewUnmarshaller creates an unmarshaller from a csv.Reader and a struct.
-func NewUnmarshaller(reader *csv.Reader, out interface{}) (*Unmarshaller, error) {
+// NewUnmarshaller is a convenience function which allocates and
+// returns a new Unmarshaller.
+func NewUnmarshaller(holder interface{}, reader Reader) (*Unmarshaller, error) {
+	return (&Config{Holder: holder}).NewUnmarshaller(reader)
+}
+
+// NewUnmarshaller creates an unmarshaller from a Reader and a struct.
+func (c *Config) NewUnmarshaller(reader Reader) (*Unmarshaller, error) {
 	headers, err := reader.Read()
 	if err != nil {
 		return nil, err
 	}
-	headers = normalizeHeaders(headers)
 
-	um := &Unmarshaller{reader: reader, outType: reflect.TypeOf(out)}
-	err = validate(um, out, headers)
+	vc, err := c.validate(headers)
 	if err != nil {
 		return nil, err
 	}
+
+	um := &Unmarshaller{
+		reader: reader,
+		config: vc,
+		line:   1,
+	}
+
 	return um, nil
 }
 
-// Read returns an interface{} whose runtime type is the same as the struct that
-// was used to create the Unmarshaller.
+// Read returns an interface{} whose runtime type is the same as the
+// struct that was used to create the Unmarshaller.
 func (um *Unmarshaller) Read() (interface{}, error) {
-	row, err := um.reader.Read()
-	if err != nil {
-		return nil, err
+	value, _, err := um.ReadUnmatched()
+	return value, err
+}
+
+func ReadAllCallback(um *Unmarshaller, cb func(interface{}) error) error {
```

Comment:
I put this in as a helper method because I wasn't convinced that the `*Unmarshaller` needed this in it, but I could be convinced to move it inside.  Gocsv had so many specialized things shoved everywhere that I think I'd rather start simple and only add stuff unless it's very warranted.

Reply:
I also hate the name.

Reply:
No.  The Java code works because Java has a type system with generics.  Since Go lacks that, it cannot express the output type as anything but `[]interface{}`, meaning any caller would need to traverse it to do something useful.

The `ReadAllCallback` function forces callers to do that as part of the parsing, which saves you from repeating the loop over the slice, and allocating two slices when you only need one.

Reply:
> it cannot express the output type as anything but []interface{}, meaning any caller would need to traverse it to do something useful.

Agreed returning a `[]interface{}` is sub-optimal, but I believe we could instead use an appropriately-typed slice, which we pass around as an `interface{}` if we:

1. Take a `holder` as an `interface{}` parameter, assert that it's a `[]T`, and append records to `holder` instead of returning them, or
2. Create our own `[]T`, to which we append records, then return as an `interface{}`.

Reply:
I like this suggestion, I'll see if I can get something reasonably working that implements it.